bin/passenger-recycler

```diff
@@ -41,41 +41,76 @@ def process_status?(pid)
 end
 
 require 'phusion_passenger'
+PhusionPassenger.locate_directories
 require 'phusion_passenger/platform_info'
 require 'phusion_passenger/platform_info/ruby'
 require 'phusion_passenger/admin_tools/memory_stats'
-PhusionPassenger.locate_directories
 stats = PhusionPassenger::AdminTools::MemoryStats.new
-unless stats.platform_provides_private_dirty_rss_information?
-  puts 'Please run as root or platform unsupported'
-  exit 1
-end
-killed = 0
-stats.passenger_processes.each do |p|
-  pid = p.pid.to_i
-  debug "Checking #{pid} with RSS of #{p.private_dirty_rss}"
-  next unless p.private_dirty_rss > CONFIG[:MAX_PRIV_RSS_MEMORY]
-  started = begin
+
+case PhusionPassenger::PlatformInfo.os_name_simple
+when "linux"
+  unless stats.platform_provides_private_dirty_rss_information?
+    puts 'Please run as root or platform unsupported'
+    exit 1
+  end
+  killed = 0
+  stats.passenger_processes.each do |p|
+    pid = p.pid.to_i
+    debug "Checking #{pid} with RSS of #{p.private_dirty_rss}"
+    next unless p.private_dirty_rss > CONFIG[:MAX_PRIV_RSS_MEMORY]
+    started = begin
               `ps -p#{pid} -o start=`.strip
             rescue StandardError => e
               verbose "Error: #{e.message} \nReturning '?'"
               '?'
             end
-  status_ps = `ps -p#{pid} -u`
-  status_all = `passenger-status 2> /dev/null`
-  status_backtraces = `passenger-status --show=backtraces 2>/dev/null`
-  verbose "Terminating #{pid} (started #{started}) with private dirty RSS" \
+    status_ps = `ps -p#{pid} -u`
+    status_all = `passenger-status 2> /dev/null`
+    status_backtraces = `passenger-status --show=backtraces 2>/dev/null`
+    verbose "Terminating #{pid} (started #{started}) with private dirty RSS" \
           " size of #{p.private_dirty_rss} MB"
-  Process.kill 'SIGUSR1', pid
-  sleep CONFIG[:GRACEFUL_SHUTDOWN_SLEEP]
-  kill(pid)
-  process_status?(pid)
-  if CONFIG[:SEND_STATUS]
-    verbose status_ps
-    verbose status_all
-    verbose status_backtraces
+    Process.kill 'SIGUSR1', pid
+    sleep CONFIG[:GRACEFUL_SHUTDOWN_SLEEP]
+    kill(pid)
+    process_status?(pid)
+    if CONFIG[:SEND_STATUS]
+      verbose status_ps
+      verbose status_all
+      verbose status_backtraces
+    end
+    killed += 1
+    exit(1) if killed >= CONFIG[:MAX_TERMINATION]
+  end
+when "freebsd"
+  killed = 0
+  stats.passenger_processes.each do |p|
+    pid = p.pid.to_i
+    pid_rss = `procstat -r #{pid} | awk '/maximum RSS/{print$5}'`.strip.to_i
```

Comment:
Could this be extracted to a ```get_pid_mem_mb``` so we can reduce the amount of duplication?

Reply:
Absolutely! I´ve also noticed that there are functions I overlooked in phusion_passenger that negates the the need to use specific shell commands, e.g: 

    pid_rss = p.rss.to_i
    pid_rss_mb = p.rss_in_mb

Which where the only variables I really needed. Beware though, that the debug and verbose messages are also different, and needs to change, since on FreeBSD, I´m not using "private dirty RSS", only RSS!

Reply:
Would you like me to make a new pull request like that instead?